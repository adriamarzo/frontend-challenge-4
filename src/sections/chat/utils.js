import Message from '../../models/message-model';

export function parseMessages(messages) {
  return Object.values(messages).reduce((obj, next) => {
    obj[next.key] = new Message(next).toJSON();
    return obj;
  }, {});
}

export function messagesFilter(data) {
  /** Support for basic message types */
  return ['text', 'dialog'].includes(data.type);
}

export function scrollBottom(container) {
  if (!container) {
    return;
  }

  container.scrollTo({
    top: container.scrollHeight,
    behavior: 'smooth',
  });
}
