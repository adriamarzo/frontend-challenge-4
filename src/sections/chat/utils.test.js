import { messagesFilter, parseMessages, scrollBottom } from './utils';

describe('Chat utils', () => {
  describe('messagesFilter', () => {
    it('should return true when the message type is supported', () => {
      expect(messagesFilter({ type: 'text' })).toBe(true);
      expect(messagesFilter({ type: 'dialog' })).toBe(true);
    });

    it('Should return false when the message type is not supported', () => {
      expect(messagesFilter({ type: 'unknown' })).toBe(false);
    });
  });

  describe('parseMessages', () => {
    it('should parse the messages object using the Message Model', () => {
      const messages = {
        x: {
          key: 'key',
          title: 'The title',
          samurai: true,
          timestamp: 1637533924211,
          type: 'text',
        },
        y: {
          key: 'key 2',
          title: 'The title 2',
          timestamp: 1637533924221,
          type: 'text',
        },
      };
      expect(parseMessages(messages)).toMatchSnapshot();
    });
  });

  describe('scrollBottom', () => {
    it('should call scrollTo with the expected arguments', () => {
      const mockDOMElement = {
        scrollTo: jest.fn(),
        scrollHeight: 1,
      };

      scrollBottom(mockDOMElement);
      expect(mockDOMElement.scrollTo).toBeCalledWith({
        behavior: 'smooth',
        top: 1,
      });
    });
  });
});
