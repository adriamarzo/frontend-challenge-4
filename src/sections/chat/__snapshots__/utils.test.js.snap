// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat utils parseMessages should parse the messages object using the Message Model 1`] = `
Object {
  "key": Object {
    "author": "bot",
    "key": "key",
    "text": "The title",
    "timestamp": 1637533924211,
    "type": "text",
  },
  "key 2": Object {
    "author": "user",
    "key": "key 2",
    "text": "The title 2",
    "timestamp": 1637533924221,
    "type": "text",
  },
}
`;
